File: 227.basic_calculator_II/answer.dart
const Set<String> primeSet = {'+', '-','*', '/'};
const Set<String> priorityPrime = {'*', '/'};


class Solution {
  int calculate(String s) {
    List<String> removedSpaceStrings = s.replaceAll(" ", '')
        .split('').reversed
        .toList();

    List<dynamic> stack = [];

    while (removedSpaceStrings.isNotEmpty) {
      // Terms가 전부 스택으로 들어가지 않은 경우
      // Terms에서 하나씩 제거 하면서 돌아간다.

      if (primeSet.contains(removedSpaceStrings.last)) {
        // 괄호가 없으니 * / 는 즉시 계산
        if (priorityPrime.contains(removedSpaceStrings.last)) {
          stack.add(operate(stack.removeLast(),removedSpaceStrings.removeLast(),
              parseNumber(removedSpaceStrings)
          ));
          // + - 는 스택에 넣은 후 계산
        } else {
          stack.add(removedSpaceStrings.removeLast());
          if (stack.length > 3) {
            stack.insert(0, operate(stack.removeAt(0),stack.removeAt(0),stack.removeAt(0)));
          }
        }
        // 기호가 아니면 숫자로 취급
      } else {
        stack.add(parseNumber(removedSpaceStrings));
      }
    }
    while (stack.length >= 3) {
      stack.insert(
          0, operate(stack.removeAt(0),stack.removeAt(0),stack.removeAt(0)));
    }

    return stack.first;
  }


  int operate(int number1, String prime, int number2) {
    switch (prime) {
      case '+':
        return number1 + number2;
      case '-':
        return number1 - number2;
      case '/':
        return number1 ~/ number2;
      case '*':
        return number1 * number2;
      default:
        throw UnimplementedError();
    }
  }


  int parseNumber(List<String> removedSpaceStrings) {
    var accString = '';
    while (removedSpaceStrings.isNotEmpty &&
        !primeSet.contains(removedSpaceStrings.last)) {
      accString += removedSpaceStrings.removeLast();
    }

    return int.parse(accString);
  }
}

Runtime: 471 ms, faster than 100.00% of Dart online submissions for Basic Calculator II.
Memory Usage: 147.7 MB MB, less than 100.00% of Dart online submissions for Basic Calculator II.

File: 227.basic_calculator_II/answer.kt
class Solution {
    val commonPrimes: Set<Char> = setOf<Char>('+', '-')
    val priorityPrimes: Set<Char> = setOf<Char>('*', '/')
    val primes: Set<Char> = commonPrimes + priorityPrimes

    fun calculate(s: String): Int {
        var stack: MutableList<Any> = mutableListOf()

        var numberStartingIndex = 0;

        for (i: Int in 0 until s.length) {

            if (primes.contains(s.get(i))) {
                var number = s.substring(numberStartingIndex, i).trim().toInt()

                if (!stack.isEmpty()) {

                    if (priorityPrimes.contains(stack.last())) {
                        var lastPrime = stack.removeAt(stack.lastIndex)
                        var lastNumber = stack.removeAt(stack.lastIndex)
                        number = operate(lastNumber as Int, lastPrime as Char, number as Int)

                    } else if (stack.size >= 3 && commonPrimes.contains(s.get(i))) {
                        stack.add(
                            0,
                            operate(stack.removeAt(0) as Int, stack.removeAt(0) as Char, stack.removeAt(0) as Int)
                        )
                    }
                }

                stack.add(number)
                stack.add(s[i]);
                numberStartingIndex = i + 1;

            }

            if (i == s.length - 1) {
                var number = s.substring(numberStartingIndex).trim().toInt()
                if (!stack.isEmpty() && priorityPrimes.contains(stack.last())) {
                    var lastPrime = stack.removeAt(stack.lastIndex)
                    var lastNumber = stack.removeAt(stack.lastIndex)
                    number = operate(lastNumber as Int, lastPrime as Char, number as Int)
                }

                stack.add(number)
            }
        }

        while (stack.size >= 3) {
            stack.add(0, operate(stack.removeAt(0) as Int, stack.removeAt(0) as Char, stack.removeAt(0) as Int))
        }

        return stack.first() as Int
    }

    fun operate(number1: Int, prime: Char, number2: Int): Int {
        return when (prime) {
            '+' ->
                number1 + number2;
            '-' ->
                number1 - number2;
            '/' ->
                number1 / number2;
            '*' ->
                number1 * number2;
            else ->
                throw IllegalArgumentException()
        }
    }
}